### ISSUE TITLE
Basic systemd support calling sd_listen_fds

### ISSUE THREAD

**Issue by gcartier** [gambit/gambit#629]
Systemd is a Linux daemon that can amongst other things be used to listen on a specific TCP port and automatically start a server (implemented in Gambit) when that port is triggered. The server then needs to call the sd_listen_fds function to retrieve the actual socket from systemd.

**Comment by lassik** [comment]
Could a more general solution be invented? inetd and tcpserver are two other ways to start servers on demand using pre-allocated sockets. Both of them do it by spawning a new server process; I suppose systemd sends the new sockets to an existing process using `SCM_RIGHTS` or something like it. FastCGI and SimpleCGI also have something like this.

**Comment by lassik** [comment]
This might even be a good topic for a SRFI.

**Comment by feeley** [comment]
I'm not sure what level of integration with Gambit you are requesting.  My first reaction is to question the portability of `sd_listen_fds`:

1) On Unix it requires linking to an external library (libsystemd) and this goes against Gambit's design which depends on no external library.

2) `int` file descriptors are a Unix thing and Windows generally uses the `HANDLE` type... so how does this work on Windows?

Aside from the portability issue, it seems that (if the library is available) a basic interface to `sd_listen_fds` can be done in user code with a `c-lambda`:

    (define sd_listen_fds (c-lambda (int) int "sd_listen_fds"))

Moreover, if I understand correctly, the functionality of `sd_listen_fds` can also be achieved  in user code by reading the environment variables $LISTEN_FDS, $LISTEN_PID and $LISTEN_FDNAMES.

Given that the feature is rather special purpose and not essential to the Gambit RTS, this seems like the perfect situation to implement it as a Gambit module.

Can you give some code showing how this feature would be used?

**Comment by gcartier** [comment]
I agree with your comments Marc. When I created the issue, I was going through every commit in Jazz's Gambit and did not reflect on the relevance of integrating systemd support in Gambit.

Here is the code that implements this in Jazz's Gambit:

Note that I did not add any flag to open-tcp-server to ask for systemd support. It is a kind of a hack where I know that the open-tcp-server call that needs systemd is called first and only once.

My only question is can I do the same thing with user code, e.g. create a tcp server that will retrieve its socket from systemd using only c-lambda? (note that if the answer in no I can still manage with a custom commit in Jazz's Gambit)

```___SCMOBJ ___device_tcp_server_setup
   ___P((___device_tcp_server **dev,
         ___device_group *dgroup,
         struct sockaddr *local_addr,
         SOCKET_LEN_TYPE local_addrlen,
         int backlog,
         int options,
         ___tls_context *tls_context),
        (dev,
         dgroup,
         local_addr,
         local_addrlen,
         backlog,
         options,
         tls_context)
___device_tcp_server **dev;
___device_group *dgroup;
struct sockaddr *local_addr;
SOCKET_LEN_TYPE local_addrlen;
int backlog;
int options;
___tls_context *tls_context;)
{
  ___SCMOBJ e;
  SOCKET_TYPE s;
  ___device_tcp_server *d;

#ifdef USE_SYSTEMD
  static int sd_listen_done = 0;

  int n = (sd_listen_done) ? -1 : sd_listen_fds(0);
  if (n > 1)
  {
    fprintf(stderr, "Too many file descriptors received.\n");
    exit(1);
  }
  else if (n == 1)
  {
    s = SD_LISTEN_FDS_START + 0;
    set_socket_blocking_mode (s, 0);
    sd_listen_done = 1;
  }
  else
  {
#endif

  if ((e = create_socket (&s, local_addr, local_addrlen, options))
      != ___FIX(___NO_ERR))
    return e;

  if (set_socket_blocking_mode (s, 0) != 0 || /* set nonblocking mode */
      listen (s, backlog) != 0)
    {
      e = ERR_CODE_FROM_SOCKET_CALL;
      CLOSE_SOCKET(s); /* ignore error */
      return e;
    }

#ifdef USE_SYSTEMD
  }
#endif
```

**Comment by feeley** [comment]
Gambit tries hard to abstract from file descriptors (which are OS dependent).  But there is a way if you are on Unix with the procedure `##open-predefined`:

    (##open-predefined (macro-direction-inout) 'name fd settings)

For example:

    > (define stdin (##open-predefined 1 'stdin 0 '())) ;; 1 = input, 0 = stdin   
    > (read stdin)
    123 ;; entered this
    123
    > stdin
    #<input-port #2 stdin>

So if you can get hold of the file descriptor in the Scheme code just call `##open-predefined`.

Of course this won't work on Windows, but neither will `sd_listen_fds`.
